src/Microsoft.Performance.SDK.Runtime.NetCoreApp/Plugins/PluginsLoader.cs

```diff
@@ -0,0 +1,255 @@
+        ///     If the caller must know which plugin(s) failed, it must use <see cref="TryLoadPlugins(IEnumerable{string}, out IList{string})"/>.
+        ///     <para/>
+        ///     This method does not report an error if a different version of one of the custom data
+        ///     sources being loaded has already been loaded by either the current or a previous call to <see cref="TryLoadPlugin(string)"/>
+        ///     or <see cref="TryLoadPlugins(IEnumerable{string}, out IList{string})"/>.
+        /// </remarks>
+        public bool TryLoadPlugin(string directory)
+        {
+            return this.TryLoadPlugins(new List<string> { directory }, out _);
+        }
+
+        /// <summary>
+        ///     Attempts to load the plugin(s) contained within any of the given directories.
+        /// </summary>
+        /// <param name="directories">
+        ///     The directories to attempt to load plugins from.
+        /// </param>
+        /// <param name="failed">
+        ///     The directories from <paramref name="directories"/> which contained at least
+        ///     one plugin that failed to load. If this method returns <c>true</c>, this
+        ///     will be empty.
+        /// </param>
+        /// <returns>
+        ///     <c>true</c> if all plugins within the given directories were loaded, <c>false</c> otherwise.
+        /// </returns>
+        /// <remarks>
+        ///     A directory from <paramref name="directories"/> will be included in <paramref name="failed"/>
+        ///     if _any_ of the plugins inside the directory fail to load. See remarks on <see cref="TryLoadPlugin(string)"/>
+        ///     for more information.
+        ///     <para/>
+        ///     This method does not report an error if a different version of one of the custom data
+        ///     sources being loaded has already been loaded by either the current or a previous call to <see cref="TryLoadPlugin(string)"/>
+        ///     or <see cref="TryLoadPlugins(IEnumerable{string}, out IList{string})"/>.
+        /// </remarks>
+        public bool TryLoadPlugins(IEnumerable<string> directories, out IList<string> failed)
+        {
+            lock (this.mutex)
+            {
+                failed = new List<string>();
+                var oldPlugins = new HashSet<CustomDataSourceReference>(this.catalog.PlugIns);
+                foreach (var dir in directories)
+                {
+
+                    if (!this.extensionDiscovery.ProcessAssemblies(dir))
+                    {
+                        failed.Add(dir);
+                        continue;
+                    }
+                }
+
+                // TODO: this does redundant work re-finalizing custom data sources
+                // loaded by previous calls to this method. The extension repository
+                // should get refactored to avoid this redundant work.
+                this.extensionRepository.FinalizeDataExtensions();
```

Comment:
if we're allowing plug-ins to be loaded at different times, then the extension repository and related code should be updated accordingly. the scenario that I'm thinking of is that a plug-in that has failed to load previously because of missing dependencies might work now. it's been a while, but I believe this could be solved fairly easily by just creating a new DataExtensionDependencyState in each DataExtensionReference where the DataExtensionDependencyState.Availability value is not Undertermined or Available. as an optimization, we might be able to just reset if Availability is MissingRequirement or MissingIndirectRequirement, but we'd need to double check that.

Reply:
Is this something that should be done in this PR? This class is primarily a wrapper around the assembly processing, so I'm not sure those changes would make sense for this PR.

FWIW, prior to this class, the SDK "allows" plugins to be loaded at different times; it just doesn't make it easy to do so.

Reply:
I created #25 as a TODO to deal with this. For reference, the concern of a plugin failing to load initially but then working later because dependencies become available may not be something we even want to deal with. Ideally a plugin includes all of its dependencies that it needs for its custom data sources to work. This scenario really could only happen if plugins are loaded through something like `-addsearchdir` but the user forgot to copy its dependencies to the folder passed in, and then a plugin later miraciously loads the dependency the original one needed. It's possible, but raises the question of how much we want to support loading plugins like that.